gateway/src/auth.rs

```diff
@@ -0,0 +1,151 @@
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+
+use rand::Rng;
+
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+use std::io::Write;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use axum::extract::{FromRequest, Extension, RequestParts, TypedHeader, Path};
+use axum::headers::{authorization::Basic, Authorization};
+use axum::response::{IntoResponse, Response};
+use std::str::FromStr;
+use axum::http::StatusCode;
+use axum::Json;
+
+use crate::service::GatewayService;
+use crate::{Error, ErrorKind, ProjectName, AccountName};
+
+#[derive(Clone, Debug, sqlx::Type, PartialEq, Hash, Eq, Serialize, Deserialize)]
+#[serde(transparent)]
+#[sqlx(transparent)]
+pub struct Key(pub String);
+
+#[async_trait]
+impl<B> FromRequest<B> for Key
+where
+    B: Send
+{
+    type Rejection = Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        TypedHeader::<Authorization<Basic>>::from_request(req)
+            .await
+            .map_err(|_| Error::from(ErrorKind::KeyMissing))
+            .and_then(|TypedHeader(Authorization(basic))| basic.password().parse())
+    }
+}
+
+impl FromStr for Key {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self(s.to_string()))
+    }
+}
+
+impl Key {
+    pub fn new_random() -> Self {
+        Self(
+            rand::thread_rng()
+                .sample_iter(&rand::distributions::Alphanumeric)
+                .take(16)
+                .map(char::from)
+                .collect::<String>(),
```

Comment:
Will need to `use rand::distributions::DistString` but looks a bit prettier :)
```suggestion
            rand::distributions::Alphanumeric.sample_string(&mut rand::thread_rng(), 16)
```